// @dojo/framework/stores/StoreInjector.mjs
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import WeakMap from '../shim/WeakMap';
import { WidgetBase } from '../core/WidgetBase';
import { handleDecorator } from '../core/decorators/handleDecorator';
import { beforeProperties } from '../core/decorators/beforeProperties';
import { alwaysRender } from '../core/decorators/alwaysRender';
import { Registry } from '../core/Registry';
import { w } from '../core/vdom';
const registeredInjectorsMap = new WeakMap();
/**
 * Decorator that registers a store injector with a container based on paths when provided
 *
 * @param config Configuration of the store injector
 */
export function storeInject(config) {
    const { name, paths, getProperties } = config;
    return handleDecorator((target, propertyKey) => {
        beforeProperties(function (properties) {
            const injectorItem = this.registry.getInjector(name);
            if (injectorItem) {
                const { injector } = injectorItem;
                const store = injector();
                const registeredInjectors = registeredInjectorsMap.get(this) || [];
                if (registeredInjectors.length === 0) {
                    registeredInjectorsMap.set(this, registeredInjectors);
                }
                if (registeredInjectors.indexOf(injectorItem) === -1) {
                    if (paths) {
                        let invalidatePaths;
                        if (typeof paths === 'function') {
                            invalidatePaths = paths(store.path);
                        }
                        else {
                            invalidatePaths = paths.map((path) => store.path(path.join('/')));
                        }
                        const handle = store.onChange(invalidatePaths, () => this.invalidate());
                        this.own({
                            destroy: () => {
                                handle.remove();
                            }
                        });
                    }
                    else {
                        this.own(store.on('invalidate', () => {
                            this.invalidate();
                        }));
                    }
                    registeredInjectors.push(injectorItem);
                }
                return getProperties(store, properties);
            }
        })(target);
    });
}
export function StoreContainer(component, name, { paths, getProperties }) {
    let WidgetContainer = class WidgetContainer extends WidgetBase {
        render() {
            return w(component, this.properties, this.children);
        }
    };
    WidgetContainer = __decorate([
        alwaysRender(),
        storeInject({ name, paths, getProperties })
    ], WidgetContainer);
    return WidgetContainer;
}
/**
 * Creates a typed `StoreContainer` for State generic.
 */
export function createStoreContainer() {
    return (component, name, { paths, getProperties }) => {
        return StoreContainer(component, name, { paths, getProperties });
    };
}
export function registerStoreInjector(store, options = {}) {
    const { key = 'state', registry = new Registry() } = options;
    if (registry.hasInjector(key)) {
        throw new Error(`Store has already been defined for key ${key.toString()}`);
    }
    registry.defineInjector(key, () => {
        return () => store;
    });
    return registry;
}
