// @dojo/framework/shim/object.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "../core/has"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var has_1 = require("../core/has");
    if (!has_1.default('es6-object')) {
        var keys_1 = Object.keys.bind(Object);
        Object.keys = function symbolAwareKeys(o) {
            return keys_1(o).filter(function (key) { return !Boolean(key.match(/^@@.+/)); });
        };
        Object.assign = function assign(target) {
            var sources = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                sources[_i - 1] = arguments[_i];
            }
            if (target == null) {
                // TypeError if undefined or null
                throw new TypeError('Cannot convert undefined or null to object');
            }
            var to = Object(target);
            sources.forEach(function (nextSource) {
                if (nextSource) {
                    // Skip over if undefined or null
                    keys_1(nextSource).forEach(function (nextKey) {
                        to[nextKey] = nextSource[nextKey];
                    });
                }
            });
            return to;
        };
        var getOwnPropertyNames_1 = Object.getOwnPropertyNames.bind(Object);
        Object.getOwnPropertyNames = function symbolAwareGetOwnPropertyNames(o) {
            return getOwnPropertyNames_1(o).filter(function (key) { return !Boolean(key.match(/^@@.+/)); });
        };
        Object.getOwnPropertySymbols = function getOwnPropertySymbols(o) {
            return getOwnPropertyNames_1(o)
                .filter(function (key) { return Boolean(key.match(/^@@.+/)); })
                .map(function (key) { return Symbol.for(key.substring(2)); });
        };
        Object.is = function is(value1, value2) {
            if (value1 === value2) {
                return value1 !== 0 || 1 / value1 === 1 / value2; // -0
            }
            return value1 !== value1 && value2 !== value2; // NaN
        };
    }
    if (!has_1.default('es2017-object')) {
        Object.getOwnPropertyDescriptors = function getOwnPropertyDescriptors(o) {
            return Object.getOwnPropertyNames(o).reduce(function (previous, key) {
                previous[key] = Object.getOwnPropertyDescriptor(o, key);
                return previous;
            }, {});
        };
        Object.entries = function entries(o) {
            return exports.keys(o).map(function (key) { return [key, o[key]]; });
        };
        Object.values = function values(o) {
            return exports.keys(o).map(function (key) { return o[key]; });
        };
    }
    exports.assign = Object.assign;
    exports.getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;
    exports.getOwnPropertyNames = Object.getOwnPropertyNames;
    exports.getOwnPropertySymbols = Object.getOwnPropertySymbols;
    exports.is = Object.is;
    exports.keys = Object.keys;
    exports.getOwnPropertyDescriptors = Object.getOwnPropertyDescriptors;
    exports.entries = Object.entries;
    exports.values = Object.values;
    exports.default = Object;
});
