// @dojo/framework/core/registerCustomElement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "./Registry", "./WidgetBase", "./vdom", "../shim/array", "../shim/global", "./mixins/Themed", "./decorators/alwaysRender"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    var Registry_1 = require("./Registry");
    var WidgetBase_1 = require("./WidgetBase");
    var vdom_1 = require("./vdom");
    var array_1 = require("../shim/array");
    var global_1 = require("../shim/global");
    var Themed_1 = require("./mixins/Themed");
    var alwaysRender_1 = require("./decorators/alwaysRender");
    var RESERVED_PROPS = ['focus'];
    var CustomElementChildType;
    (function (CustomElementChildType) {
        CustomElementChildType["DOJO"] = "DOJO";
        CustomElementChildType["NODE"] = "NODE";
        CustomElementChildType["TEXT"] = "TEXT";
    })(CustomElementChildType = exports.CustomElementChildType || (exports.CustomElementChildType = {}));
    function isElement(item) {
        return item && item.nodeType === 1;
    }
    function isDojoChild(item) {
        return isElement(item) && item.tagName.indexOf('-') > -1;
    }
    function DomToWidgetWrapper(domNode) {
        var DomToWidgetWrapper = /** @class */ (function (_super) {
            tslib_1.__extends(DomToWidgetWrapper, _super);
            function DomToWidgetWrapper() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            DomToWidgetWrapper.prototype.render = function () {
                var _this = this;
                var properties = Object.keys(this.properties).reduce(function (props, key) {
                    var value = _this.properties[key];
                    if (key.indexOf('on') === 0 || RESERVED_PROPS.indexOf(key) !== -1) {
                        key = "__" + key;
                    }
                    props[key] = value;
                    return props;
                }, {});
                return vdom_1.dom({ node: domNode, props: properties, diffType: 'dom' });
            };
            Object.defineProperty(DomToWidgetWrapper, "domNode", {
                get: function () {
                    return domNode;
                },
                enumerable: true,
                configurable: true
            });
            DomToWidgetWrapper = tslib_1.__decorate([
                alwaysRender_1.alwaysRender()
            ], DomToWidgetWrapper);
            return DomToWidgetWrapper;
        }(WidgetBase_1.WidgetBase));
        return DomToWidgetWrapper;
    }
    exports.DomToWidgetWrapper = DomToWidgetWrapper;
    function create(descriptor, WidgetConstructor) {
        var _a = descriptor.attributes, attributes = _a === void 0 ? [] : _a, _b = descriptor.registryFactory, registryFactory = _b === void 0 ? function () { return new Registry_1.default(); } : _b;
        var attributeMap = {};
        attributes.forEach(function (propertyName) {
            var attributeName = propertyName.toLowerCase();
            attributeMap[attributeName] = propertyName;
        });
        return /** @class */ (function (_super) {
            tslib_1.__extends(class_1, _super);
            function class_1() {
                var _this = _super !== null && _super.apply(this, arguments) || this;
                _this._properties = {};
                _this._children = [];
                _this._eventProperties = {};
                _this._propertiesMap = {};
                _this._initialised = false;
                _this._childType = descriptor.childType;
                return _this;
            }
            class_1.prototype.connectedCallback = function () {
                if (this._initialised) {
                    return;
                }
                this._waitTillReady();
            };
            class_1.prototype._hasBeenParsed = function () {
                if (document.readyState !== 'loading') {
                    return true;
                }
                var element = this;
                while (element) {
                    if (element.nextSibling) {
                        return true;
                    }
                    element = element.parentNode;
                }
                return false;
            };
            class_1.prototype._waitTillReady = function () {
                var _this = this;
                this._initialised = true;
                if (this._hasBeenParsed()) {
                    this._readyCallback();
                }
                else {
                    setTimeout(function () {
                        _this._waitTillReady();
                    }, 100);
                }
            };
            class_1.prototype._readyCallback = function () {
                var _this = this;
                var domProperties = {};
                var _a = descriptor.properties, properties = _a === void 0 ? [] : _a, _b = descriptor.events, events = _b === void 0 ? [] : _b;
                this._properties = tslib_1.__assign({}, this._properties, this._attributesToProperties(attributes));
                tslib_1.__spread(attributes, properties).forEach(function (propertyName) {
                    var isReservedProp = RESERVED_PROPS.indexOf(propertyName) !== -1;
                    var value = _this._propertiesMap[propertyName] || !isReservedProp ? _this[propertyName] : undefined;
                    var filteredPropertyName = propertyName.replace(/^on/, '__');
                    if (isReservedProp) {
                        filteredPropertyName = "__" + propertyName;
                    }
                    if (value !== undefined) {
                        _this._properties[propertyName] = value;
                    }
                    if (filteredPropertyName !== propertyName) {
                        domProperties[filteredPropertyName] = {
                            get: function () { return _this._getProperty(propertyName); },
                            set: function (value) { return _this._setProperty(propertyName, value); }
                        };
                    }
                    if (!isReservedProp) {
                        domProperties[propertyName] = {
                            get: function () { return _this._getProperty(propertyName); },
                            set: function (value) { return _this._setProperty(propertyName, value); }
                        };
                    }
                });
                events.forEach(function (propertyName) {
                    var eventName = propertyName.replace(/^on/, '').toLowerCase();
                    var filteredPropertyName = propertyName.replace(/^on/, '__on');
                    domProperties[filteredPropertyName] = {
                        get: function () { return _this._getEventProperty(propertyName); },
                        set: function (value) { return _this._setEventProperty(propertyName, value); }
                    };
                    _this._eventProperties[propertyName] = undefined;
                    _this._properties[propertyName] = function () {
                        var args = [];
                        for (var _i = 0; _i < arguments.length; _i++) {
                            args[_i] = arguments[_i];
                        }
                        var eventCallback = _this._getEventProperty(propertyName);
                        if (typeof eventCallback === 'function') {
                            eventCallback.apply(void 0, tslib_1.__spread(args));
                        }
                        _this.dispatchEvent(new CustomEvent(eventName, {
                            bubbles: false,
                            detail: args
                        }));
                    };
                });
                Object.defineProperties(this, domProperties);
                var children = array_1.from(this.childNodes).filter(function (childNode) { return !vdom_1.isTextNode(childNode) || childNode.data.replace(/^\s+|\s+$/g, ''); });
                if (!this._childType) {
                    if (children.some(function (child) { return isDojoChild(child); })) {
                        this._childType = CustomElementChildType.DOJO;
                    }
                    else {
                        this._childType = CustomElementChildType.NODE;
                    }
                }
                array_1.from(children).forEach(function (childNode) {
                    if (_this._childType === CustomElementChildType.DOJO) {
                        childNode.addEventListener('dojo-ce-render', function () { return _this._render(); });
                        childNode.addEventListener('dojo-ce-connected', function () { return _this._render(); });
                        _this._children.push(DomToWidgetWrapper(childNode));
                    }
                    else {
                        _this._children.push(vdom_1.dom({ node: childNode, diffType: 'dom' }));
                    }
                });
                this.addEventListener('dojo-ce-connected', function (e) { return _this._childConnected(e); });
                var widgetProperties = this._properties;
                var renderChildren = function () { return _this.__children__(); };
                var Wrapper = /** @class */ (function (_super) {
                    tslib_1.__extends(class_2, _super);
                    function class_2() {
                        return _super !== null && _super.apply(this, arguments) || this;
                    }
                    class_2.prototype.render = function () {
                        return vdom_1.w(WidgetConstructor, widgetProperties, renderChildren());
                    };
                    return class_2;
                }(WidgetBase_1.WidgetBase));
                var registry = registryFactory();
                var themeContext = Themed_1.registerThemeInjector(this._getTheme(), registry);
                global_1.default.addEventListener('dojo-theme-set', function () { return themeContext.set(_this._getTheme()); });
                var r = vdom_1.renderer(function () { return vdom_1.w(Wrapper, {}); });
                this._renderer = r;
                r.mount({ domNode: this, merge: false, registry: registry });
                var root = this.children[0];
                if (root) {
                    var _c = global_1.default.getComputedStyle(root).display, display = _c === void 0 ? 'block' : _c;
                    this.style.display = display;
                }
                this.dispatchEvent(new CustomEvent('dojo-ce-connected', {
                    bubbles: true,
                    detail: this
                }));
            };
            class_1.prototype._getTheme = function () {
                if (global_1.default && global_1.default.dojoce && global_1.default.dojoce.theme) {
                    return global_1.default.dojoce.themes[global_1.default.dojoce.theme];
                }
            };
            class_1.prototype._childConnected = function (e) {
                var _this = this;
                var node = e.detail;
                if (node.parentNode === this) {
                    var exists = this._children.some(function (child) { return child.domNode === node; });
                    if (!exists) {
                        node.addEventListener('dojo-ce-render', function () { return _this._render(); });
                        this._children.push(DomToWidgetWrapper(node));
                        this._render();
                    }
                }
            };
            class_1.prototype._render = function () {
                if (this._renderer) {
                    this._renderer.invalidate();
                    this.dispatchEvent(new CustomEvent('dojo-ce-render', {
                        bubbles: false,
                        detail: this
                    }));
                }
            };
            class_1.prototype.__properties__ = function () {
                return tslib_1.__assign({}, this._properties, this._eventProperties);
            };
            class_1.prototype.__children__ = function () {
                if (this._childType === CustomElementChildType.DOJO) {
                    return this._children.filter(function (Child) { return Child.domNode.isWidget; }).map(function (Child) {
                        var domNode = Child.domNode;
                        return vdom_1.w(Child, tslib_1.__assign({}, domNode.__properties__()), tslib_1.__spread(domNode.__children__()));
                    });
                }
                else {
                    return this._children;
                }
            };
            class_1.prototype.attributeChangedCallback = function (name, oldValue, value) {
                var propertyName = attributeMap[name];
                this._setProperty(propertyName, value);
            };
            class_1.prototype._setEventProperty = function (propertyName, value) {
                this._eventProperties[propertyName] = value;
            };
            class_1.prototype._getEventProperty = function (propertyName) {
                return this._eventProperties[propertyName];
            };
            class_1.prototype._setProperty = function (propertyName, value) {
                this._properties[propertyName] = value;
                this._render();
            };
            class_1.prototype._getProperty = function (propertyName) {
                return this._properties[propertyName];
            };
            class_1.prototype._attributesToProperties = function (attributes) {
                var _this = this;
                return attributes.reduce(function (properties, propertyName) {
                    var attributeName = propertyName.toLowerCase();
                    var value = _this.getAttribute(attributeName);
                    if (value !== null) {
                        properties[propertyName] = value;
                    }
                    return properties;
                }, {});
            };
            Object.defineProperty(class_1, "observedAttributes", {
                get: function () {
                    return Object.keys(attributeMap);
                },
                enumerable: true,
                configurable: true
            });
            Object.defineProperty(class_1.prototype, "isWidget", {
                get: function () {
                    return true;
                },
                enumerable: true,
                configurable: true
            });
            class_1.prototype.set = function (key, value) {
                this._propertiesMap[key] = value;
                if (this._renderer) {
                    this._setProperty(key, value);
                }
            };
            return class_1;
        }(HTMLElement));
    }
    exports.create = create;
    function register(WidgetConstructor) {
        var descriptor = WidgetConstructor.__customElementDescriptor;
        if (!descriptor) {
            throw new Error('Cannot get descriptor for Custom Element, have you added the @customElement decorator to your Widget?');
        }
        global_1.default.customElements.define(descriptor.tagName, create(descriptor, WidgetConstructor));
    }
    exports.register = register;
    exports.default = register;
});
